### PR TITLE
chore(stdlib): Improve examples in `Array` module

### PR DESCRIPTION
**Pull request by spotandjake**
This pull request adds examples to every function in the array module, along with standardizing the style of the examples.

I plan on going through and adding examples to all the modules in the stdlib starting with this so it makes sense to note some style decisions:
* All the examples are runnable
  *  The only thing that needs to be added is the module header and the include
  * for single line examples an assert needs to be added as well
* Single line examples are treated as if they were assertions
* Syntax for creating the data type is added as an example to the `module`, It seems like a good place to show a basic example of how to create the data type that were working with. I think this will be more helpful with the number libs later on.
* Examples are placed between `@throws` and `@since`
* Examples are mostly kept as simple as possible

### CODE REVIEW

File: stdlib/array.gr [hunk 1]
```diff
@@ -393,7 +416,7 @@ provide let reduce = (fn, initial, array) => {
  * @param array: The array to iterate
  * @returns The final accumulator returned from `fn`
  *
- * @example Array.reduceRight((a, b) => b ++ a, "", [> "baz", "bar", "foo"]) // "foobarbaz"
+ * @example Array.reduceRight((el, acc) => acc ++ el, "", [> "baz", "bar", "foo"]) == "bazbarfoo"
```
Comment (alex-snezhko) [inline stdlib/array.gr:419]: Should be "foobarbaz"

File: stdlib/array.gr [hunk 2]
```diff
@@ -440,6 +470,8 @@ provide let reducei = (fn, initial, array) => {
  *
  * @throws InvalidArgument(String): When the combined length of all arrays is not an integer
  *
+ * @example Array.flatMap(e => [> 1], [> 1, 2, 3]) == [> 1, 1, 1 ]
```
Comment (alex-snezhko) [inline stdlib/array.gr:473]: I think a better example could be used here

File: stdlib/array.gr [hunk 3]
```diff
@@ -263,6 +264,8 @@ provide let concat = arrays => {
  * @param array: The array to copy
  * @returns The new array containing the elements from the input
  *
+ * @example Array.copy([> 1, 2, 3 ]) == [> 1, 2, 3 ]
```
Comment (alex-snezhko) [inline stdlib/array.gr:267]: I think you should format the arrays and lists in these examples like how the formatter would

File: stdlib/array.gr [hunk 4]
```diff
@@ -724,6 +794,9 @@ provide let count = (fn, array) => {
  * @param array: The array to iterate
  * @returns The total number of elements that satisfy the condition
  *
+ * @example Array.counti((e, i) => e % 2 == 0 && i % 2 == 0, [> 1, 2, 3, 4, 5 ]) == 0
+ * @example Array.counti((e, i) => e % 2 == 0 && i % 2 == 0, [> 0, 1, 2, 3, 5 ]) == 1
```
Comment (alex-snezhko) [inline stdlib/array.gr:798]: 2

File: stdlib/array.gr [hunk 5]
```diff
@@ -984,10 +1079,16 @@ provide let sort = (comp, array) => {
  * @param n: The number of elements to rotate by
  * @param arr: The array to be rotated
  *
- * @example let array = [> 1, 2, 3, 4, 5]; rotate(2, arr); arr == [> 3, 4, 5, 1, 2]
- * @example let array = [> 1, 2, 3, 4, 5]; rotate(-1, arr); arr == [> 5, 1, 2, 3, 4]
- * @since v0.4.5
+ * @example
+ * let array = [> 1, 2, 3, 4, 5]
+ * Array.rotate(2, arr)
```
Comment (alex-snezhko) [inline stdlib/array.gr:1084]: `arr` -> `array`

File: stdlib/array.gr [hunk 6]
```diff
@@ -1159,6 +1264,8 @@ provide module Immutable {
   /**
    * An empty array.
    *
+   * @example Array.Immutable.empty()
```
Comment (alex-snezhko) [inline stdlib/array.gr:1267]: Not a function; also a better example may be `Array.Immutable.empty == Array.Immutable.fromList([])`

File: stdlib/array.gr [hunk 7]
```diff
@@ -1072,6 +1173,10 @@ provide let chunk = (chunkSize, arr) => {
 /**
  * An immutable array implementation.
  *
+ * @example from Array use { module Immutable }
+ * @example Array.Immutable.empty()
```
Comment (alex-snezhko) [inline stdlib/array.gr:1177]: Not a function

File: stdlib/array.gr [hunk 8]
```diff
@@ -1280,9 +1402,14 @@ provide module Immutable {
...
    * @returns A new array containing the new element at the given index
+   *
    * @throws IndexOutOfBounds: When the index being updated is outside the array's bounds
    *
-   * @example set(1, 9, fromList([1, 2, 3, 4, 5])) == fromList([1, 9, 3, 4, 5])
+   * @example
+   * from Array use { module Immutable }
+   * let mut arr = fromList([1, 2, 3, 4, 5])
+   * arr = Immutable.set(1, 9, arr)
+   * assert arr == Immutable.fromList([1, 9, 3, 4, 5])
```
Comment (alex-snezhko) [inline stdlib/array.gr:1412]: May be useful to assert that the original is unchanged

File: stdlib/array.gr [hunk 9]
```diff
@@ -1454,7 +1581,11 @@ provide module Immutable {
    * @param array2: The array containing elements to appear second
    * @returns The new array containing elements from `array1` followed by elements from `array2`
    *
-   * @example append(fromList([1, 2]), fromList([3, 4, 5])) == fromList([1, 2, 3, 4, 5])
+   * @example
+   * from Array use { module Immutable }
+   * let arr1 = Immutable.fromList([1, 2])
+   * let arr2 = Immutable.fromList([3, 4, 5])
+   * assert Immutable.append(arr1, arr2) == Immutable.fromList([1, 2, 3, 4, 5])
```
Comment (alex-snezhko) [inline stdlib/array.gr:1588]: Likewise to above comment, may be useful to assert original unchanged

File: stdlib/array.gr [hunk 10]
```diff
@@ -1579,6 +1729,13 @@ provide module Immutable {
    * @param n: The number of times to iterate the given array
    * @param array: The array to iterate
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr = Immutable.fromList(["a", "b", "c"])
+   * let mut str = ""
+   * Immutable.forEach(e => str = str ++ e, 2, arr)
```
Comment (alex-snezhko) [inline stdlib/array.gr:1736]: `forEach` -> `cycle`

File: stdlib/array.gr [hunk 11]
```diff
@@ -1763,6 +1953,15 @@ provide module Immutable {
    * @param array: The array to check
    * @returns `true` if all elements satify the condition or `false` otherwise
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr = Immutable.fromList(['a', 'a' ])
```
Comment (alex-snezhko) [inline stdlib/array.gr:1958]: Nit remove extra space

File: stdlib/array.gr [hunk 12]
```diff
@@ -1779,6 +1978,15 @@ provide module Immutable {
    * @param array: The array to iterate
    * @returns `true` if one or more elements satify the condition or `false` otherwise
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr = Immutable.fromList(['a', 'b' ])
```
Comment (alex-snezhko) [inline stdlib/array.gr:1983]: Remove extra space

File: stdlib/array.gr [hunk 13]
```diff
@@ -1859,6 +2100,12 @@ provide module Immutable {
    * @param array2: The array to provide values for the second tuple element
    * @returns The new array containing all pairs of `(a, b)`
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr1 = Immutable.fromList([1, 2])
+   * let arr2 = Immutable.fromList([3, 4])
+   * assert Immutable.product(arr1, arr2) == Immutable.fromList([ (1, 3), (1, 4), (2, 3), (2, 4)])
```
Comment (alex-snezhko) [inline stdlib/array.gr:2107]: Remove extra space

File: stdlib/array.gr [hunk 14]
```diff
@@ -1878,6 +2125,11 @@ provide module Immutable {
    * @param array: The array to iterate
    * @returns The total number of elements that satisfy the condition
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr = Immutable.fromList([1,1,2,3,4])
```
Comment (alex-snezhko) [inline stdlib/array.gr:2130]: Fix spacing

File: stdlib/array.gr [hunk 15]
```diff
@@ -1893,6 +2145,11 @@ provide module Immutable {
    * @param array: The array to filter
    * @returns The new array with only unique values
    *
+   * @example
+   * from Array use { module Immutable }
+   * let arr = Immutable.fromList([1,1,2,3,2,4])
```
Comment (alex-snezhko) [inline stdlib/array.gr:2150]: Fix spacing

File: stdlib/array.gr [hunk 16]
```diff
@@ -1937,8 +2200,16 @@ provide module Immutable {
...
+   * @example
+   * from Array use { module Immutable }
+   * let arr1 = Immutable.fromList([1, 2, 3])
+   * let arr2 = Immutable.fromList([4, 5, 6])
+   * assert Immutable.zipWith((a, b) => a + b, arr1, arr2) == Immutable.fromList([5, 7, 9])
+   * @example
+   * from Array use { module Immutable }
+   * let arr1 = Immutable.fromList([1, 2, 3])
+   * let arr2 = Immutable.fromList([4, 5, 6])
+   * assert Immutable.zipWith((a, b) => a * b, arr1, arr2) == Immutable.fromList([4, 10])
```
Comment (alex-snezhko) [inline stdlib/array.gr:2212]: [4, 10, 18]

**Comment by spotandjake** [comment]
Made those changes

**Review by alex-snezhko — APPROVED** [review]

**Review by phated — APPROVED** [review]

**Pull request merged**